### ISSUE TITLE
[REVIEW]: PyPVRPM: Photovoltaic Reliability and Performance Model in Python

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#4093]
**Submitting author:** @brandons209 (<a href="http://orcid.org/0000-0001-8278-7158">Brandon Silva</a>)
**Repository:** https://github.com/FSEC-Photovoltaics/pvrpm-lcoe
**Branch with paper.md** (empty if default branch): 
**Version:** 1.7.4
**Editor:** @jgostick
**Reviewers:** @blthayer, @SarthakJariwala
**Archive:** 10.5281/zenodo.6355115

 

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/359bb7f809f245df315e45d3e7bb9213/status.svg)](https://joss.theoj.org/papers/359bb7f809f245df315e45d3e7bb9213)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/359bb7f809f245df315e45d3e7bb9213"><img src="https://joss.theoj.org/papers/359bb7f809f245df315e45d3e7bb9213/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/359bb7f809f245df315e45d3e7bb9213/status.svg)](https://joss.theoj.org/papers/359bb7f809f245df315e45d3e7bb9213)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/FSEC-Photovoltaics/pvrpm-lcoe" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@SarthakJariwala, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @jgostick know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @SarthakJariwala
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/FSEC-Photovoltaics/pvrpm-lcoe">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/FSEC-Photovoltaics/pvrpm-lcoe">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@brandons209) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] #4213

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @blthayer
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/FSEC-Photovoltaics/pvrpm-lcoe">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/FSEC-Photovoltaics/pvrpm-lcoe">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@brandons209) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @SarthakJariwala it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Wordcount for `paper.md` is 728

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.22 s (384.1 files/s, 165021.5 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
JavaScript                      14           2405           2473           9211
HTML                            24           3699             72           8933
SVG                              1              0              0           2671
Python                          16            746            840           2645
CSS                              4            191             35            757
YAML                             6            113            116            683
JSON                             5              0              0            496
reStructuredText                11            179            312            180
TeX                              1              7              0             65
Markdown                         2             17              0             43
DOS Batch                        1              8              1             26
make                             1              4              7              9
-------------------------------------------------------------------------------
SUM:                            86           7369           3856          25719
-------------------------------------------------------------------------------


Statistical information for the repository '318eb3e546f0763db97f5a72' was
gathered on 2022/01/22.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Brandon                         25          2729           1089           17.80
Brandon Silva                    9         15281            152           71.95
brandons209                     15          1876            324           10.26

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Brandon                   18320          671.3          2.9               16.18
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.2172/1088066 is OK
- 10.2172/1367461 is OK
- 10.1109/PVSC.2018.8547454 is OK
- 10.1109/PVSC.2017.8366397 is OK
- 10.2172/1761998 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.04093/joss.04093/10.21105.joss.04093.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.04093/joss.04093/10.21105.joss.04093.pdf) :page_facing_up: :point_left:

**Comment by jgostick** [comment]
@whedon add @blthayer as reviewer

**Comment by whedon** [comment]
OK, @blthayer is now a reviewer

**Comment by jgostick** [comment]
Hi @blthayer and @SarthakJariwala.  Thank you for being a reviewer for JOSS.  At the top of this thread you will see a checklist for you to use as a guideline for your review.  At JOSS we generally work with the authors to help them get their submission into publishable form (if it is not already), so you can interact with them here or on their repo regarding any changes you think are needed.  Remember that the review is publish and non-anonymous, so is meant to be a collaborative effort, and your github handles will appear on the final PDF as reviewers, so you'll get credit for your efforts. 

Let's aim to have this done in a about 1 month?  So Feb 28th?

**Comment by blthayer** [comment]
FYI (@brandons209, @jgostick): At this point I plan to start my review in earnest either this weekend or next week. Looking forward to checking out `PyPVRPM`!

**Comment by brandons209** [comment]
@blthayer Great! If you need any help or want some example cases for testing let me know.

**Comment by whedon** [comment]
:wave: @SarthakJariwala, please update us on how your review is going (this is an automated reminder).

**Comment by blthayer** [comment]
@jgostick - it looks like I need a fresh invitation sent in order to be able to edit my checklist. When I went to the following URL, it said the invitation was expired: https://github.com/openjournals/joss-reviews/invitations

This isn't (yet) blocking me from performing my review.

Thanks!

**Comment by blthayer** [comment]
I'll use this comment to maintain a checklist of the issues which I've filed over in the author's repository.

- [ ] https://github.com/FSEC-Photovoltaics/pvrpm-lcoe/issues/27
- [ ] https://github.com/FSEC-Photovoltaics/pvrpm-lcoe/issues/28
- [ ] https://github.com/FSEC-Photovoltaics/pvrpm-lcoe/issues/29
- [ ] https://github.com/FSEC-Photovoltaics/pvrpm-lcoe/issues/30
- [ ] https://github.com/FSEC-Photovoltaics/pvrpm-lcoe/issues/31
- [ ] https://github.com/FSEC-Photovoltaics/pvrpm-lcoe/issues/32
- [ ] (OPTIONAL) https://github.com/FSEC-Photovoltaics/pvrpm-lcoe/issues/33
- [ ] https://github.com/FSEC-Photovoltaics/pvrpm-lcoe/issues/34
- [ ] (OPTIONAL) https://github.com/FSEC-Photovoltaics/pvrpm-lcoe/issues/35
- [ ] https://github.com/FSEC-Photovoltaics/pvrpm-lcoe/issues/36

**Comment by blthayer** [comment]
My review is currently blocked by https://github.com/FSEC-Photovoltaics/pvrpm-lcoe/issues/36, @brandons209

**Comment by jgostick** [comment]
@whedon assign @blthayer as reviewer

**Comment by whedon** [comment]
OK, @blthayer is now a reviewer

**Comment by jgostick** [comment]
@whedon add @SarthakJariwala as reviewer

**Comment by whedon** [comment]
OK, @SarthakJariwala is now a reviewer

**Comment by SarthakJariwala** [comment]
@brandons209 I will use this comment as a checklist for all the issues I have filled in your repository.

- [ ] https://github.com/FSEC-Photovoltaics/pvrpm-lcoe/issues/49
- [ ] https://github.com/FSEC-Photovoltaics/pvrpm-lcoe/issues/50
- [ ] https://github.com/FSEC-Photovoltaics/pvrpm-lcoe/issues/51
- [ ] https://github.com/FSEC-Photovoltaics/pvrpm-lcoe/issues/52

**Comment by jgostick** [comment]
I have pinged both reviewers by email to see how things are progressing.

**Comment by blthayer** [comment]
@brandons209 - FYI, I wasn't able to find time to work on my review this past weekend (Feb 26 and 27). I'll make another pass this upcoming weekend (Mar 05 and 06). My apologies for the delay.

**Comment by blthayer** [comment]
@brandons209 - We're close! I've added three issues (one optional) to my [issue-tracking-checklist](https://github.com/openjournals/joss-reviews/issues/4093#issuecomment-1030686522). After these are addressed, I will consider my review complete.

Nice work! I hope that between mine and Sarthak's review that you feel that `PyPVRPM` has been made better and more approachable.

My general advice for you going forward:

Adding a real-world case study to your documentation could be useful for a newcomer understanding the what and why of `PyPVRPM`. For example, something like "PyPVRPM was used to analyze a PV installation with a capacity of X to evaluate if using infrared drone monitoring was a cost effective preventative maintenance strategy. It was found that Y more MWh would be produced over the plant's lifetime for a total cost savings of Z."

In my experience, good software development practices aren't something taught in school (at least not in electrical engineering), and these skills are something you have to pick up and hone on your own. I didn't read the source code in depth, but I did glance at your testing suite. For starters, I want to say it's _excellent_ that `a)` you have tests and that `b)` you have a full end-to-end simulation test. However, I found the unit tests to be lacking. Learning to write good tests and testable software takes time, but it's well worth the effort. Consider trying to break the functionality of `PyPVRPM` into smaller, easily testable pieces. Using a tool like [pytest-cov](https://pypi.org/project/pytest-cov/) can help you find untested areas of your software (always shoot for 100% coverage! :smile: ). Once you have a comprehensive test suite put together, you can feel really confident that the addition of a new feature or a bug fix didn't break anything or cause regressions. Additionally, rigorous testing helps ensure that not only does the software run, but that it is _correct_. All too often I see academic research software with little to no tests, and in my opinion, untested software is incorrect, untrustworthy software. Alright, putting my testing soap box away now :smile: 

I hope that industry and researchers alike pick up `PyPVRPM` and find it useful!

@jgostick - once I'm satisfied that Mr. Silva has addressed https://github.com/FSEC-Photovoltaics/pvrpm-lcoe/issues/58 and https://github.com/FSEC-Photovoltaics/pvrpm-lcoe/issues/60, I'll consider my review complete, and will say as much in a new comment in this thread.

**Comment by brandons209** [comment]
> Adding a real-world case study to your documentation could be useful for a newcomer understanding the what and why of PyPVRPM. For example, something like "PyPVRPM was used to analyze a PV installation with a capacity of X to evaluate if using infrared drone monitoring was a cost effective preventative maintenance strategy. It was found that Y more MWh would be produced over the plant's lifetime for a total cost savings of Z."

We are planning to add a real world case, once we finish up a paper that uses PyPVRPM for showing how monitoring effects LCOE. The paper should be done in the next month or two (hopefully).

>In my experience, good software development practices aren't something taught in school (at least not in electrical engineering), and these skills are something you have to pick up and hone on your own. I didn't read the source code in depth, but I did glance at your testing suite. ...

Thank you for the kind words! I do agree that in academia documentation and good software practices aren't used that often like for industry products (and even a lot of industry products too!). 
I also agree that comment about testing, but for PVRPM I was having trouble figuring out what to add for unit tests. Everything is tested with the simulation test, so I wasn't sure what else to add. I'll check out that package though!

**Comment by blthayer** [comment]
@brandons209 - thanks for addressing the issues, and good work!

@jgostick - I now consider my review complete. ✔️ 

Please let me know if there's anything needed from me in order to take this submission over the finish line.

**Comment by jgostick** [comment]
Hi @brandons209, below are a few checkboxes for you to address so we can move ahead with this:

# Additional Author Tasks After Review is Complete
- [ ] Double check authors and affiliations (including ORCIDs)
- [ ] Make a release of the software with the latest changes from the review and post the version number here. This is the version that will be used in the JOSS paper.  
- [ ] Archive the release on Zenodo/figshare/etc and post the DOI here.
- [ ] Make sure that the title and author list (including ORCIDs) in the archive match those in the JOSS paper.

**Comment by jgostick** [comment]
# Editor Task Prior to Acceptance
- [ ] Set article DOI with ``@whedon set <DOI here> as archive``
- [ ] Set version with ``@whedon set <version here> as version``
- [ ] Double check rendering of paper with ``@whedon generate pdf``
- [ ] Specifically check the references with ``@whedon check references`` and ask author(s) to update as needed
- [ ] Do a 'dry run' of acceptance with ``@whedon recommend accept``

**Comment by jgostick** [comment]
@whedon check references

**Comment by jgostick** [comment]
@editorialbot check references

**Comment by jgostick** [comment]
@editorialbot generate pdf

**Comment by SarthakJariwala** [comment]
> Ok, there is one checkbox left for @SarthakJariwala, but I think we can overlook based on @blthayer deep review. So, let's proceed with acceptance! I will do it this afternoon.

@jgostick I haven't been able to verify the functionality of the software primarily because it isn't running as expected on macOS. Otherwise, I am happy with everything else and agree you can overlook that given @blthayer review of the functionality.

@brandons209 I would recommend you add a note for your users in your docs (preferably in the installation section) that there are issues getting it to work on macOS. Based on our discussion in https://github.com/FSEC-Photovoltaics/pvrpm-lcoe/issues/49 I agree it is likely a PySAM issue. Great work and good luck!

**Comment by brandons209** [comment]
Version 1.7.4 is the final release version for the paper. All the authors are correct.

Here is the archive on Zenodo: [10.5281/zenodo.6355115](https://doi.org/10.5281/zenodo.6355115)

I have confirmed the title and everything else looks good for the paper!

**Comment by jgostick** [comment]
@editorialbot set 1.7.4 as the version

**Comment by jgostick** [comment]
@editorialbot set 10.5281/zenodo.6355115 as archive

**Comment by jgostick** [comment]
@editorialbot recommend accept

**Comment by jgostick** [comment]
@editorialbot commands

**Comment by jgostick** [comment]
@editorialbot recommend-accept

**Comment by jgostick** [comment]
@brandons209 Congrats.  An editor in chief will be notified that the submission is ready to go, and will make it official shortly.

**Comment by kthyng** [comment]
@jgostick Is there a reason one checkbox is unchecked?

@brandons209 Please edit the metadata in your Zenodo archive so that the title and author list match your JOSS paper. Also it looks like the version listed here and in the Zenodo archive are not the same, so please rectify that and let me know what version you want here.

**Comment by danielskatz** [comment]
@brandons209 - I'll proofread the paper now (as the AEiC on duty this week).

**Comment by danielskatz** [comment]
@brandons209 - I'm suggesting some small language changes in https://github.com/FSEC-Photovoltaics/pvrpm-lcoe/pull/61 - please merge this or let me know what you disagree with.  Once we do this, and get the last checkbox issue resolve, we can move to final acceptance and publication

**Comment by danielskatz** [comment]
@jgostick - thanks - I missed that earlier discussion.

**Comment by brandons209** [comment]
@danielskatz Looks good, I merged the PR.

**Comment by danielskatz** [comment]
@editorialbot recommend-accept

**Comment by danielskatz** [comment]
@editorialbot accept

**Comment by danielskatz** [comment]
Congratulations to @brandons209 and co-authors!!

And thanks to @blthayer and @SarthakJariwala for reviewing, and @jgostick for editing!
We couldn't do this without you
